File: src/pal/openai_provider.py
"""
OpenAI provider implementation.

Supports GPT-4, GPT-3.5, and compatible APIs.
"""

import time
from typing import Any, Dict, Iterator, List, Optional

import httpx

from pal.base import BaseProvider, ProviderCapabilities, ProviderFactory


class OpenAIProvider(BaseProvider):
    """
    Provider for OpenAI API.
    
    Supports GPT-4, GPT-3.5-turbo, and compatible endpoints.
    """
    
    DEFAULT_BASE_URL = "https://api.openai.com/v1"
    
    def __init__(self, config: Dict[str, Any]):
        super().__init__(config)
        self.base_url = config.get("base_url", self.DEFAULT_BASE_URL)
        self.api_key = config.get("api_key") or self._get_api_key(config)
        
        if not self.api_key:
            raise ValueError("OpenAI API key not provided. Set OPENAI_API_KEY env var or api_key in config.")
        
        # HTTP client
        self.client = httpx.Client(
            base_url=self.base_url,
            headers={
                "Authorization": f"Bearer {self.api_key}",
                "Content-Type": "application/json",
            },
            timeout=self.timeout,
        )
        
        # Pricing per 1K tokens (as of 2024)
        self.pricing = {
            "gpt-4": {"input": 0.03, "output": 0.06},
            "gpt-4-turbo": {"input": 0.01, "output": 0.03},
            "gpt-3.5-turbo": {"input": 0.0005, "output": 0.0015},
        }
    
    def _get_api_key(self, config: Dict[str, Any]) -> Optional[str]:
        """Get API key from environment or config."""
        import os
        env_var = config.get("api_key_env", "OPENAI_API_KEY")
        return os.environ.get(env_var)
    
    def generate(
        self,
        messages: List[Dict[str, str]],
        params: Optional[Dict[str, Any]] = None
    ) -> str:
        """
        Generate text using OpenAI chat completion API.
        """
        params = params or {}
        
        payload = {
            "model": self.model,
            "messages": messages,
            "temperature": params.get("temperature", 0.7),
            "max_tokens": params.get("max_tokens", 2000),
            "top_p": params.get("top_p", 0.9),
            "stream": False,
        }
        
        # Add JSON mode if requested
        if params.get("json_mode"):
            payload["response_format"] = {"type": "json_object"}
        
        try:
            response = self.client.post("/chat/completions", json=payload)
            response.raise_for_status()
            
            data = response.json()
            return data["choices"][0]["message"]["content"]
        
        except httpx.HTTPStatusError as e:
            if e.response.status_code == 401:
                raise RuntimeError("Invalid OpenAI API key")
            elif e.response.status_code == 429:
                raise RuntimeError("OpenAI rate limit exceeded")
            raise RuntimeError(f"OpenAI API error: {e}")
    
    def generate_stream(
        self,
        messages: List[Dict[str, str]],
        params: Optional[Dict[str, Any]] = None
    ) -> Iterator[str]:
        """
        Generate text with streaming.
        """
        params = params or {}
        
        payload = {
            "model": self.model,
            "messages": messages,
            "temperature": params.get("temperature", 0.7),
            "max_tokens": params.get("max_tokens", 2000),
            "stream": True,
        }
        
        try:
            with self.client.stream("POST", "/chat/completions", json=payload) as response:
                response.raise_for_status()
                
                for line in response.iter_lines():
                    if line and line.startswith("data: "):
                        data_str = line[6:]  # Remove "data: " prefix
                        
                        if data_str == "[DONE]":
                            break
                        
                        try:
                            import json
                            data = json.loads(data_str)
                            delta = data["choices"][0].get("delta", {})
                            content = delta.get("content", "")
                            if content:
                                yield content
                        except (json.JSONDecodeError, KeyError):
                            continue
        
        except httpx.HTTPError as e:
            raise RuntimeError(f"OpenAI streaming error: {e}")
    
    def capabilities(self) -> ProviderCapabilities:
        """Report OpenAI capabilities."""
        ctx_len = 128000 if "gpt-4-turbo" in self.model else (
            8192 if "gpt-4" in self.model else 16385
        )
        
        return ProviderCapabilities(
            ctx_len=ctx_len,
            supports_tools=True,
            supports_json_mode=True,
            supports_thinking_mode=False,
            supports_streaming=True,
        )
    
    def healthcheck(self) -> bool:
        """Check if API is accessible."""
        try:
            response = self.client.get("/models", timeout=10)
            return response.status_code == 200
        except httpx.RequestError:
            return False
    
    def price_estimate(self, input_tokens: int, output_tokens: int) -> float:
        """
        Estimate cost for generation.
        
        Returns cost in USD.
        """
        model_pricing = self.pricing.get(self.model, self.pricing.get("gpt-3.5-turbo"))
        
        input_cost = (input_tokens / 1000) * model_pricing["input"]
        output_cost = (output_tokens / 1000) * model_pricing["output"]
        
        return round(input_cost + output_cost, 4)


# Register provider
ProviderFactory.register("openai", OpenAIProvider)

File: go/pkg/agents/openai_provider.go
package agents

import (
	"bytes"
	"context"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"os"
	"strings"
	"time"

	"github.com/novelist/novelist/pkg/models"
)

const defaultOpenAIBaseURL = "https://api.openai.com/v1"

type openAIProvider struct {
	model      string
	baseURL    string
	apiKey     string
	client     *http.Client
	chatPath   string
	modelsPath string
}

type openAIRequest struct {
	Model          string    `json:"model"`
	Messages       []Message `json:"messages"`
	Temperature    float64   `json:"temperature,omitempty"`
	MaxTokens      int       `json:"max_tokens,omitempty"`
	TopP           float64   `json:"top_p,omitempty"`
	ResponseFormat any       `json:"response_format,omitempty"`
}

type openAIResponse struct {
	Choices []struct {
		Message struct {
			Content string `json:"content"`
		} `json:"message"`
	} `json:"choices"`
	Usage struct {
		PromptTokens     int `json:"prompt_tokens"`
		CompletionTokens int `json:"completion_tokens"`
	} `json:"usage"`
	Error *struct {
		Message string `json:"message"`
	} `json:"error,omitempty"`
}

func init() {
	RegisterProviderFactory("openai", NewOpenAIProvider)
}

// NewOpenAIProvider creates a provider backed by OpenAI-compatible API.
func NewOpenAIProvider(config models.ProviderConfig) (Provider, error) {
	if strings.TrimSpace(config.Model) == "" {
		return nil, fmt.Errorf("openai provider requires model")
	}

	baseURL := strings.TrimRight(strings.TrimSpace(config.BaseURL), "/")
	if baseURL == "" {
		baseURL = defaultOpenAIBaseURL
	}

	apiKeyEnv := strings.TrimSpace(config.APIKeyEnv)
	if apiKeyEnv == "" {
		apiKeyEnv = "OPENAI_API_KEY"
	}
	apiKey := strings.TrimSpace(os.Getenv(apiKeyEnv))
	if apiKey == "" {
		return nil, fmt.Errorf("openai API key is missing in env %s", apiKeyEnv)
	}

	timeout := time.Duration(config.Timeout) * time.Second
	if timeout <= 0 {
		timeout = 60 * time.Second
	}

	chatPath, modelsPath := openAIPaths(baseURL)

	return &openAIProvider{
		model:      config.Model,
		baseURL:    baseURL,
		apiKey:     apiKey,
		client:     &http.Client{Timeout: timeout},
		chatPath:   chatPath,
		modelsPath: modelsPath,
	}, nil
}

func (p *openAIProvider) Name() string {
	return "openai"
}

func (p *openAIProvider) Generate(ctx context.Context, messages []Message, params GenerateParams) (*models.GenerationResult, error) {
	payload := openAIRequest{
		Model:       p.model,
		Messages:    messages,
		Temperature: params.Temperature,
		MaxTokens:   params.MaxTokens,
		TopP:        params.TopP,
	}
	if params.JSONMode {
		payload.ResponseFormat = map[string]string{"type": "json_object"}
	}

	body, err := json.Marshal(payload)
	if err != nil {
		return nil, fmt.Errorf("failed to marshal openai request: %w", err)
	}

	req, err := http.NewRequestWithContext(ctx, http.MethodPost, p.baseURL+p.chatPath, bytes.NewReader(body))
	if err != nil {
		return nil, fmt.Errorf("failed to build openai request: %w", err)
	}
	req.Header.Set("Content-Type", "application/json")
	req.Header.Set("Authorization", "Bearer "+p.apiKey)

	resp, err := p.client.Do(req)
	if err != nil {
		return nil, fmt.Errorf("openai request failed: %w", err)
	}
	defer resp.Body.Close()

	raw, err := io.ReadAll(io.LimitReader(resp.Body, 2*1024*1024))
	if err != nil {
		return nil, fmt.Errorf("failed reading openai response: %w", err)
	}

	var out openAIResponse
	if err := json.Unmarshal(raw, &out); err != nil {
		return nil, fmt.Errorf("failed to decode openai response: %w", err)
	}

	if resp.StatusCode >= http.StatusBadRequest {
		msg := strings.TrimSpace(string(raw))
		if out.Error != nil && out.Error.Message != "" {
			msg = out.Error.Message
		}
		return nil, fmt.Errorf("openai returned %d: %s", resp.StatusCode, msg)
	}

	if len(out.Choices) == 0 {
		return nil, fmt.Errorf("openai response missing choices")
	}

	text := strings.TrimSpace(out.Choices[0].Message.Content)
	promptTokens := out.Usage.PromptTokens
	completionTokens := out.Usage.CompletionTokens
	if promptTokens <= 0 {
		promptTokens = estimateTokensFromMessages(messages)
	}
	if completionTokens <= 0 {
		completionTokens = estimateTokensFromText(text)
	}

	return &models.GenerationResult{
		Text:             text,
		PromptTokens:     promptTokens,
		CompletionTokens: completionTokens,
	}, nil
}

func (p *openAIProvider) Capabilities() ProviderCapabilities {
	ctxLen := 16385
	if strings.Contains(p.model, "gpt-4") {
		ctxLen = 8192
	}
	if strings.Contains(p.model, "gpt-4o") || strings.Contains(p.model, "gpt-4.1") {
		ctxLen = 128000
	}

	return ProviderCapabilities{
		CtxLen:            ctxLen,
		SupportsTools:     true,
		SupportsJSONMode:  true,
		SupportsThinking:  false,
		SupportsStreaming: false,
	}
}

func (p *openAIProvider) HealthCheck(ctx context.Context) error {
	req, err := http.NewRequestWithContext(ctx, http.MethodGet, p.baseURL+p.modelsPath, nil)
	if err != nil {
		return err
	}
	req.Header.Set("Authorization", "Bearer "+p.apiKey)

	resp, err := p.client.Do(req)
	if err != nil {
		return err
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		raw, _ := io.ReadAll(io.LimitReader(resp.Body, 2048))
		return fmt.Errorf("status %d: %s", resp.StatusCode, strings.TrimSpace(string(raw)))
	}
	return nil
}

func openAIPaths(baseURL string) (chatPath string, modelsPath string) {
	if strings.HasSuffix(baseURL, "/v1") {
		return "/chat/completions", "/models"
	}
	return "/v1/chat/completions", "/v1/models"
}
